crates/ruff_linter/src/rules/pylint/rules/redefined_slots_in_subclass.rs

```diff
@@ -0,0 +1,215 @@
+use std::hash::Hash;
+
+use ruff_python_semantic::{analyze::class::any_super_class, SemanticModel};
+use rustc_hash::FxHashSet;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for a re-defined slot in a subclass.
+///
+/// ## Why is this bad?
+/// If a class defines a slot also defined in a base class, the
+/// instance variable defined by the base class slot is inaccessible
+/// (except by retrieving its descriptor directly from the base class).
+///
+/// ## Example
+/// ```python
+/// class Base:
+///     __slots__ = ("a", "b")
+///
+///
+/// class Subclass(Base):
+///     __slots__ = ("a", "d")  # [redefined-slots-in-subclass]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Base:
+///     __slots__ = ("a", "b")
+///
+///
+/// class Subclass(Base):
+///     __slots__ = "d"
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct RedefinedSlotsInSubclass {
+    name: String,
+}
+
+impl Violation for RedefinedSlotsInSubclass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedefinedSlotsInSubclass { name } = self;
+        format!("Redefined slots ['{name}'] in subclass")
```

Comment:
Not sure if we need the brackets, you'll also need to update the snapshots
```suggestion
        format!("Redefined slots '{name}' in subclass")
```

Additionally, I think we can improve the message by including the name of the subclass in which Ruff found the original definition. This can be done in a follow-up PR as I think it'll involve either updating `any_super_class` or adding a `iter_super_class` function which iterates over all the classes.